### ISSUE TITLE
Vice theme sidebar won't load

### ISSUE THREAD

**Issue by nickl1** [comfusion/hyperdrive#7]
hello, i'm using the premium Vice wordpress theme and when i enable hyperdrive the sidebar refuses to load when i click on it.

**Comment by jhabdas** [comment]
Thank you for trying the Hyperdrive beta @nickl1. The vice theme looks wonderful, and I have a [special place in my heart](https://github.com/jhabdas/lumpen-radio) for music. Would you be so kind as to share the following:

Operating System
Browser
Link to a live demo of the premium Vice theme you're running, if available (I saw two different looking Vice themes during my search).

Having this information will be helpful in debugging the issue and release a solid 1.0.0 plugin for the open source community.

**Comment by monolux** [comment]
@nickl1 Is [this](https://themeforest.net/item/vice-music-band-dj-and-radio-wordpress-theme/10067870) the Vice theme you are talking about?

@jhabdas Josh, dear Josh, I cannot but say it otherwise, [fetch/inject](https://github.com/jhabdas/fetch-inject) is **the shiznit** and I cannot wait to get my hands dirty using this.

Just developing a WP theme with performance in mind, and I mean **real** performance that goes all the way down to how to have your HTML arranged and how deep one has CSS selectors, [design decisions](http://designingforperformance.com/), etc. etc. and to close the lip on all this fetch/inject will come in super handy! Great stuff!

The use cases on [the main libs page](https://github.com/jhabdas/fetch-inject) give me goosebumps in a positive sense, if you like I will, with a bit of work, create a few demos where people new to this could see a real difference, possibly even hosting the demos somewhere so people can paste the before and after/with and without links to PageSpeed and see with their own eyes.

It is stuff like this, easily available to newcomers that will eventually make the web a better place for all, by giving easy to understand access to technologies that really every website **should have**.

**Comment by nickl1** [comment]
server is debian, my pc is windows 8.1, chrome. could i email you the website somehow? the public vice demo is unavailable.

@monolux this is it.

**Comment by jhabdas** [comment]
Will look into this soon. Not a top priority until 1.0.0 milestones are complete.

**Comment by nickl1** [comment]
i'd guess this is a general problem, not specifically with that particular wp theme

**Comment by jhabdas** [comment]
@nick1 I appreciate and respect your concerns. Could you provide your current version of Chrome and use the Chrome Developer Tools Network panel to capture, zip and attach here a HAR file containing the page with all assets after it is fully loaded with Hyperdrive enabled? Having those will be useful in determining the nature and cause of your issue.

Thank you for your help.

**Comment by nickl1** [comment]
attaching a zip doens't work for some reason here, can i have your email so i can email it to you or send you a link to the website?

**Comment by jhabdas** [comment]
jhabdas@gmail.com

**Comment by jhabdas** [comment]
@nickl1 took a look at your HAR file, made some observations and have some thoughts. will respond soon with more info and observations. and may also reach out to the theme maintainer to get a copy of the theme. before I do, could you tell me which version of the theme you're using, if available?

**Comment by nickl1** [comment]
it's Version: 22.5.9, i guess the latest one

**Comment by jhabdas** [comment]
Great. If you have a direct connection to the maintainer of the theme, please shoot me their email address privately. If not, that's okay too. Cheers.

**Comment by jhabdas** [comment]
@nickl1 Sorry for the delay. Due to its complexity, Vice is going to make a great theme for us to use as a test case as we work through any bugs found and wrap up the beta. For vice I see the _dark matter_ injected into your hyperdrive isn't quite right. I'm going to talk to engineering to see what we can do about it.

**Bridge to engineering. This is the captain. Were you able to debug the dark matter clogging up the Vice hyperdrive?**

_momentary pause followed by cool noise_

**Engineering: We believe we've isolated the problem. The Vice galaxy is exhibiting what appears to be a molecular destabilization causing a temporal shift in our thrusters, over.**

_more cool noises_

**Bridge: Thanks for the update. Report back as soon as you have a solution.**

---

Dumping debug data.

```
fetchInject([
  "",
  "/wp-includes/js/wp-embed.min.js",
  "http://website.com/wp-content/plugins/contact-form-7/includes/js/scripts.js?ver=4.7",
  "http://website.com/wp-content/plugins/qt-swipebox/swipebox/js/jquery.swipebox.min.js",
  "http://website.com/wp-content/plugins/revslider/public/assets/js/jquery.themepunch.revolution.min.js?ver=5.2.6",
  "http://website.com/wp-content/plugins/revslider/public/assets/js/jquery.themepunch.tools.min.js?ver=5.2.6",
  "http://website.com/wp-content/themes/vice/js/jquery.fullPage.js?ver=20140825",
  "http://website.com/wp-content/themes/vice/js/jquery.slimScroll.js?ver=20140825",
  "http://website.com/wp-content/themes/vice/js/qt-jquerylibraries.js?ver=20140825",
  "http://website.com/wp-content/themes/vice/js/qt-main.js?ver=20140825",
  "https://maps.googleapis.com/maps/api/js",
  "https://www.google.com/jsapi"
], fetchInject([
  "",
  "http://website.com/wp-content/themes/vice/js/jquery.fullPage.js?ver=20140825",
  "http://website.com/wp-content/themes/vice/js/jquery.slimScroll.js?ver=20140825",
  "http://website.com/wp-content/themes/vice/js/qt-jquerylibraries.js?ver=20140825",
  "https://maps.googleapis.com/maps/api/js"
], fetchInject([
  "",
  "http://website.com/wp-content/themes/vice/js/jquery.fullPage.js?ver=20140825",
  "http://website.com/wp-content/themes/vice/js/jquery.slimScroll.js?ver=20140825"
], fetchInject([
  "",
  "http://website.com/wp-content/themes/vice/js/jquery.slimScroll.js?ver=20140825"
], fetchInject([
  "/wp-includes/js/jquery/jquery-migrate.min.js?ver=1.4.1",
  "/wp-includes/js/jquery/jquery.js?ver=1.12.4"
])))));
```

It appears we're trying to load blank strings. We'll need to take a look at that.

---

**Comment by monolux** [comment]
@jhabdas If only a fraction of the devs could keep an easy humorous approach.. having a good laugh following this issue, **not** for its content but for its from, you rock man :)

**Comment by nickl1** [comment]
hello, i found out that the plugin will also bug with the drop down menu of the free [poseidon theme](https://themezee.com/themes/poseidon/).. :(

**Comment by jhabdas** [comment]
@nickl1 see https://github.com/wp-id/hyperdrive/issues/16 for my thoughts on Poseidon theme. A little verbose but I don't want to have to field another issue like that (which IMHO is a theme issue we may be able to improve upon but not before 1.0.0 as it can be covered in documentation or resolved by the theme maintainer).

As for Vice, I cannot further debug the issue until I get a hold of the source code. Because we're building a free plugin here I hope you will perceive it as reasonable I will not be paying to license a copy of Vice to debug this issue. That said, if you'd like to foot the bill I'm happy to debug the issue to completion (or until this issue is closed later for other reasons).

**Comment by jhabdas** [comment]
@nickl1 You can log a support request for the Vice theme here: http://www.qantumthemes.com/helpdesk/forums/forum/vice-music-dj-and-music-band-wordpress-theme/

If they can get me a copy of their theme I will tell them what needs to be changed to make it compatible with Hyperdrive. As it stands I can see from the debug dump above something may be wrong with the way they're enqueuing scripts. Please feel free to reopen this if you have more information allowing me to reliably reproduce or debug the issue. Thanks!

**Comment by jhabdas** [comment]
The issues shown in the [above debug dump](https://github.com/wp-id/hyperdrive/issues/7#issuecomment-298728177) have been resolved in [`1.0.0-beta.2`](https://github.com/wp-id/hyperdrive/blob/master/CHANGELOG.md#100-beta2-2017-05-09). It's possible Vice may work now. Unconfirmed as I do not have a copy of the paid theme for testing purposes.

**Comment by jhabdas** [comment]
Poseidon sidebar issue resolved in Beta 6. Vice theme should work now too. As well as 90% of all of the themes out there. Here's the line of code that did it:

```php
\add_action( 'wp_enqueue_scripts', __NAMESPACE__ . '\engage', 1987 )
```

Once I understood `_WP_Dependency`, which is extended by `WP_Scripts` and `WP_Styles` is really a state machine the approach to solve became rather clear:

Simply hook into the same place where themes and plugins register their scripts and styles, and then engage at a very low priority (in this case, the year the film Spaceballs was released). This approach should prevent the majority theme and plugin maintainers from needing to adjust any of their work. I've started a general discussion thread where I'm documenting minimum support requirements for Hyperdrive here: #70.

Here's what Poseidon's new waterfall looks like prior to #75, which will further increase the speed. These metrics were taken by Chrome Dev tools emulating a Good 4G connection using stock Poseidon theme and with an unprimed browser cache:

<img width="526" alt="screen shot 2017-06-10 at 3 54 19 pm" src="https://user-images.githubusercontent.com/440298/27001157-3be1332e-4df5-11e7-8496-2a1aec65bfe4.png">

**Comment by nickl1** [comment]
unfortunately it still doesn't work with beta6

**Comment by jhabdas** [comment]
Could you send me another HAR dump or expose the site so I can prod at it (I think I know which site it is, even though you took are to remove PII from the last HAR emailed)? Without that info fixing this is like building a ship inside a bottle with a blindfold on.

**Comment by jhabdas** [comment]
> unfortunately it still doesn't work with beta6

And by "it", I'm assuming you're referring specifically to Vice and not Poseidon.

**Comment by nickl1** [comment]
oh, the poseidon drop down is working now. i've emailed you the vice site

**Comment by jhabdas** [comment]
@nickl1 got it. looking in just a moment. could you activate the plugin in 5 minutes and ping me once activated so I can start the probe?

**Comment by nickl1** [comment]
i activated it

**Comment by nickl1** [comment]
activated it 

    On Saturday, June 10, 2017 2:03 PM, Josh Habdas <notifications@github.com> wrote:
 

 @nickl1 got it. looking in just a moment. could you activate the plugin in 5 minutes and ping me once activated so I can start the probe?—

**Comment by jhabdas** [comment]
@nickl1 captured HAR. please deactivate for comparison.

**Comment by nickl1** [comment]
done, mind i'm using cloudflare, i don't know if this will affect things

**Comment by jhabdas** [comment]
CloudFlare is cool. No problems there. Since the early betas I added a version string to the output to help me debug. Right now I see the following in the HAR: `Hyperdrive v1.0.0-beta.5`. Let me look at the release to be sure I cut beta 6 properly. Get back to you shortly (may need to eat quickly in between).

**Comment by nickl1** [comment]
i have installed hyperdrive-1.0.0-beta.6, which has the wrong number inside hyperdrive.php " * Version:         1.0.0-beta.5"

**Comment by jhabdas** [comment]
You sure did. I derped the release. Let me see if it has the right code in it.

**Comment by jhabdas** [comment]
Okay. Looks like the goods are in place, but the version string was missing in the tagged update due to a sloppy interactive rebase (doing a bit of this manually right now until automated). Going to parse over the payloads in the HAR and eat. Get back to you soon with findings. At first glance I can see jQuery is coming down via Fetch so that's promising as it'll already speed up the site once we get the other issues looked at.

If possible, a quick summary of what you're not seeing you expect to see will help. But looking at the theme site with Hyperdrive enabled I can clearly see the background videos may not be loading correctly over XHR (which is what Fetch replaces, and the core technology Hyperdrive leverages to function).

**Comment by nickl1** [comment]
the sidebar won't open when hyperdrive is enabled

**Comment by jhabdas** [comment]
@nickl1 could you dig into the theme files and see if how it's hooking into the action api? I'm curious if they're using `wp_enqueue_script` or something else.

**Comment by jhabdas** [comment]
Root cause of sidebar issue in Vice theme identified in a reduced test case whereby Vice was tested with Hyperdrive in isolation.

The issue is caused by the following two lines of code in Vice's `functions.php` file:

    wp_enqueue_script('google-maps',$mapsurl);$dependency[] = 'google-maps';
    wp_enqueue_script('google-jsapi','https://www.google.com/jsapi');

When one attempts to make a `fetch` request to a server which does not provide the appropriate CORS headers to allow cross-origin resource sharing the browser issue the following response:

```
Fetch API cannot load https://maps.googleapis.com/maps/api/js. No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

This has the effect of preventing the browser from executing `qt-main.js`, which uses jQuery to register a `click` listener on the sidebar hamburger anchor element.

If the two above lines of code are commented out the `fetch` response is allowed to process and the sidebar works as expected (under the reduced test case).

Given Hyperdrive cannot control the response headers web servers provide in response to asynchronous requests I need to mull on this for a few to consider how best to address this particular scenario.

Meanwhile, please enjoy this very pretty waterfall produced by throttling Vice with Hyperdrive to a Good 4G connection speed with Googlie stuffs disabled:

<img width="526" alt="screen shot 2017-06-11 at 11 56 25 am" src="https://user-images.githubusercontent.com/440298/27008163-28cac5ea-4e9d-11e7-83bf-8ac83c4bc292.png">

**Comment by jhabdas** [comment]
Please see #6 for bug to be used in addressing CORS headers issue.

**Comment by nickl1** [comment]
would disabling external js fetching be a viable temporary fix?

**Comment by jhabdas** [comment]
> would disabling external js fetching be a viable temporary fix?

Oh I've got [a few options](https://github.com/comfusion/hyperdrive/issues/6#issuecomment-307609498) cooking. And that's one of them. But I need time to look at the performance, maintainability and security implications and a spoon full of sugar to make the medicine go down.

**Comment by jhabdas** [comment]
Man, PHP is so weird. I'm actually getting variable results on server-side code in different browsers. I couldn't even make this shit up.

**Comment by jhabdas** [comment]
LOLOL. In Fx and Safari, if you switch the order of the following lines of code you get different result when dumping the value of `$antiparticles`; but in Chrome the result stays the same:

https://github.com/comfusion/hyperdrive/blob/v1.0.0-beta.6/src/hyperdrive.php#L117-L118

![picardpalm](https://user-images.githubusercontent.com/440298/27010472-9b3dc214-4ed7-11e7-9abc-c0b081c6d41d.jpg)

**Comment by jhabdas** [comment]
Will loop back to this after a stiff drink and some zZzzzzZZZzz

**Comment by jhabdas** [comment]
Blocked by #78, which outputs unexpected results when admin logged in.

**Comment by jhabdas** [comment]
@nickl1 believe I discovered the root cause of #78. here it is: https://youtu.be/xVxuHqmNpbI

**Comment by jhabdas** [comment]
Did some more debugging on Vice. I've provided guidance in #70 as to how to properly handle the CORS issue per Hyperdrive guidance and stated the limitation. May revisit later. But need to see more real world use cases first.

For the sidebar, it's actually a Vice issue at its core. It relates to how Vice uses Ajax to rewrite the DOM. For Vice to work with Hyperdrive the theme maintainer will need to follow my guidance in #70 and do some testing to update their theme for modern web standards. When they do it'll work just fine, as shown here:

![output](https://user-images.githubusercontent.com/440298/27083816-5c433354-507c-11e7-945b-a5d06365b560.gif)

@nickl1 Thanks for raising this issue. It helped identify some bugs which have since been corrected in Hyperdrive, providing Poseidon (and likely thousands of other themes) basic compatibility for a speed boost using Fetch Injection.
